src/mono/mono/metadata/marshal.c

```diff
@@ -6237,31 +6237,37 @@ mono_marshal_emit_native_wrapper (MonoImage *image, MonoMethodBuilder *mb, MonoM
 	get_marshal_cb ()->emit_native_wrapper (image, mb, sig, piinfo, mspecs, func, flags);
 }
 
-static MonoMarshalLightweightCallbacks marshal_lightweight_cb;
-static gboolean lightweight_cb_inited = FALSE;
+static MonoMarshalLightweightCallbacks* marshal_lightweight_cb = NULL;
+
 
 void
 mono_install_marshal_callbacks (MonoMarshalLightweightCallbacks *cb)
 {
-	g_assert (!lightweight_cb_inited);
 	g_assert (cb->version == MONO_MARSHAL_CALLBACKS_VERSION);
-	memcpy (&marshal_lightweight_cb, cb, sizeof (MonoMarshalLightweightCallbacks));
-	lightweight_cb_inited = TRUE;
+	MonoMarshalLightweightCallbacks* local_cb = (MonoMarshalLightweightCallbacks*)malloc(sizeof(MonoMarshalLightweightCallbacks));
+	memcpy (local_cb, cb, sizeof (MonoMarshalLightweightCallbacks));
+
+	if (mono_atomic_cas_ptr((void**)&marshal_lightweight_cb, local_cb, NULL))
```

Comment:
same comments as the other method:
1. use `g_malloc`/`g_free`
2. barrier after memcpy
3. add `!= NULL` to the `if` condition